src/Swarm/Web.hs

```diff
@@ -0,0 +1,41 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Swarm.Web where
+
+import Control.Lens ((^.))
+import Control.Monad.IO.Class (liftIO)
+import Data.IORef (IORef, readIORef)
+import Data.IntMap qualified as IM
+import Data.Maybe (fromMaybe)
+import Network.Wai qualified
+import Network.Wai.Handler.Warp (Port)
+import Network.Wai.Handler.Warp qualified
+import Servant
+import Swarm.Game.Robot
+import Swarm.Game.State
+
+type SwarmApi =
+  "robots" :> Get '[JSON] [Robot]
+    :<|> "robot" :> Capture "id" Int :> Get '[JSON] Robot
+
+mkApp :: IORef GameState -> Servant.Server SwarmApi
+mkApp gsRef =
+  robotsHandler
+    :<|> robotHandler
+ where
+  robotsHandler = do
+    g <- liftIO (readIORef gsRef)
+    pure $ IM.elems $ g ^. robotMap
+  robotHandler rid = do
+    g <- liftIO (readIORef gsRef)
+    pure $ fromMaybe (error "Unknown robot") (IM.lookup rid (g ^. robotMap))
```

Comment:
Currently, requesting a robot that doesn't exist generates a "connection was reset" error in the web browser, and spews a bunch of errors in the terminal.  How hard would it be to instead serve a proper error message via the web server?  That seems like it would be more reasonable, and wouldn't mess up the terminal display.

Reply:
Of course, using Maybe does the trick, and the handler is even shorter :)